### DISCUSSION TITLE
Could relay cache be shared across php cli processes?

### DISCUSSION THREAD

**Discussion by dorrogeray** [cachewerk/relay#147] [Ideas]
Would it be technically possible to share the relay cache not just between php-fpm processes, but also between php cli processes? Using something like [shared system memory](https://www.php.net/manual/en/ref.shmop.php), where some id would be provided in relay extension configuration to identify such shared memory? Therefore you could have 10 php-fpm processes sharing memory with 10 php cli workers, or have them configured to use separate memory blocks (1 for workers and 1 for php-fpm processes).

Asking because it would be really nice if cli workers could also leverage local relay cache.

**Comment by michael-grunder** [comment]
Yes, it is technically possible.

Relay uses a bespoke heap allocator (a malloc-like API analog) that operates against `mmap`'ed shared memory. To make this work with System V shared memory we would need to add a mechanism to turn offsets into the appropriate virtual memory address for each attached client (as they'll likely map to different virtual addresses).
